Source/towners.cpp

```diff
@@ -355,365 +314,296 @@ void TownerTalk(_speech_id message)
+					quests[Q_MUSHROOM]._qmsg = TEXT_MUSH9;
+					InitQTextMsg(TEXT_MUSH9);
+					return;
+				}
 			}
-		} else {
-			if (player.HasItem(IDI_SPECELIX)) {
-				witch._tbtcnt = true;
-				witch._tTalkingToPlayer = &player;
-				InitQTextMsg(TEXT_MUSH12);
-				quests[Q_MUSHROOM]._qactive = QUEST_DONE;
-				witch._tMsgSaid = true;
-				AllItemsList[IDI_SPECELIX].iUsable = true;
-			} else if (player.HasItem(IDI_BRAIN) && quests[Q_MUSHROOM]._qvar2 != TEXT_MUSH11) {
-				witch._tbtcnt = true;
-				witch._tTalkingToPlayer = &player;
-				quests[Q_MUSHROOM]._qvar2 = TEXT_MUSH11;
-				InitQTextMsg(TEXT_MUSH11);
-				witch._tMsgSaid = true;
+			if (quests[Q_MUSHROOM]._qvar1 >= QS_MUSHGIVEN) {
+				if (player.HasItem(IDI_BRAIN)) {
+					quests[Q_MUSHROOM]._qmsg = TEXT_MUSH11;
+					InitQTextMsg(TEXT_MUSH11);
+					return;
+				}
+				if (player.HasItem(IDI_SPECELIX)) {
+					InitQTextMsg(TEXT_MUSH12);
+					quests[Q_MUSHROOM]._qactive = QUEST_DONE;
+					AllItemsList[IDI_SPECELIX].iUsable = true; // TODO modefy item instead of AllItemsList
+					return;
+				}
 			}
 		}
 	}
-	if (!qtextflag) {
-		TownerTalk(TEXT_ADRIA1);
-		StartStore(STORE_WITCH);
-	}
+
+	TownerTalk(TEXT_ADRIA1);
+	StartStore(STORE_WITCH);
 }
 
-void TalkToBarmaid(PlayerStruct &player, TownerStruct &barmaid)
+void TalkToBarmaid(PlayerStruct &player, TownerStruct & /*barmaid*/)
 {
 	if (!player._pLvlVisited[21] && player.HasItem(IDI_MAPOFDOOM)) {
 		quests[Q_GRAVE]._qactive = QUEST_ACTIVE;
 		quests[Q_GRAVE]._qlog = true;
 		quests[Q_GRAVE]._qmsg = TEXT_GRAVE8;
 		InitQTextMsg(TEXT_GRAVE8);
-		barmaid._tMsgSaid = true;
-	}
-	if (!qtextflag) {
-		TownerTalk(TEXT_GILLIAN1);
-		StartStore(STORE_BARMAID);
+		return;
 	}
+
+	TownerTalk(TEXT_GILLIAN1);
+	StartStore(STORE_BARMAID);
 }
 
-void TalkToDrunk(PlayerStruct &player, TownerStruct &barmaid)
+void TalkToDrunk(PlayerStruct & /*player*/, TownerStruct & /*drunk*/)
 {
 	TownerTalk(TEXT_FARNHAM1);
 	StartStore(STORE_DRUNK);
 }
 
 void TalkToHealer(PlayerStruct &player, TownerStruct &healer)
 {
-	if (!gbIsMultiplayer) {
-		if (player._pLvlVisited[1] || (gbIsHellfire && player._pLvlVisited[5])) {
-			if (!healer._tMsgSaid) {
-				if (quests[Q_PWATER]._qactive == QUEST_INIT) {
-					quests[Q_PWATER]._qactive = QUEST_ACTIVE;
-					quests[Q_PWATER]._qlog = true;
-					quests[Q_PWATER]._qmsg = TEXT_POISON3;
-					quests[Q_PWATER]._qvar1 = 1;
-					healer._tbtcnt = true;
-					healer._tTalkingToPlayer = &player;
-					InitQTextMsg(TEXT_POISON3);
-					healer._tMsgSaid = true;
-				} else if (quests[Q_PWATER]._qactive == QUEST_DONE && quests[Q_PWATER]._qvar1 != 2) {
-					quests[Q_PWATER]._qvar1 = 2;
-					healer._tbtcnt = true;
-					healer._tTalkingToPlayer = &player;
-					InitQTextMsg(TEXT_POISON5);
-					SpawnUnique(UITEM_TRING, healer.position.x, healer.position.y + 1);
-					healer._tMsgSaid = true;
-				}
-			}
+	if (quests[Q_PWATER]._qactive != QUEST_NOTAVAIL) {
+		if (player._pLvlVisited[1] && quests[Q_PWATER]._qactive == QUEST_INIT) {

```

Comment:
```suggestion
		if (player._pLvlVisited[1] || player._pLvlVisited[5]) && quests[Q_PWATER]._qactive == QUEST_INIT) {
```
Hellfire allows you to skip level 1 so keep this.